File: cpp/dsa-cmake/src/ColaSecuencial.cpp
#include "../include/ColaSecuencial.hpp"
#include <iostream>

using namespace std;

ColaSecuencial::ColaSecuencial(int capacidad) : __capacidad(capacidad) {
  __primero = 0;
  __ultimo = 0;
  __cola = new int[capacidad];
}

void ColaSecuencial::insertar(int item) {
  if (llena()) {
    cout << "cola llena." << endl;
    return;
  }
  __cola[__ultimo++] = item;
}

int ColaSecuencial::eliminar() {
  if (vacia()) {
    cout << "cola vacia." << endl;
    return -1;
  }
  int temp = __cola[__primero];
  int i = __primero;
  while (i < __ultimo - 1) {
    __cola[i] = __cola[i + 1];
    i++;
  }
  __ultimo--;
  return temp;
}

void ColaSecuencial::recorrer() const {
  int i = __primero;
  while (i < __ultimo) {
    cout << __cola[i] << " ";
    i++;
  }
  cout << endl;
}

bool ColaSecuencial::vacia() const { return __ultimo == __primero; }

bool ColaSecuencial::llena() const { return __ultimo == __capacidad; }

File: python/repaso/cola_secuencial.py
import numpy as np


class Cola:
    __pri: int
    __ult: int
    __cap: int
    __cola: np.ndarray

    def __init__(self, cap) -> None:
        self.__cap = cap
        self.__pri = 0
        self.__ult = 0
        self.__cola = np.empty(cap, dtype=int)

    def insertar(self, item):
        if self.llena():
            print("cola llena.")
            return
        self.__cola[self.__ult] = item
        self.__ult += 1

    def eliminar(self):
        if self.vacia():
            print("cola vacia")
            return
        temp = self.__cola[self.__pri]
        i = self.__pri
        while i < self.__ult - 1:
            self.__cola[i] = self.__cola[i + 1]
            i += 1
        self.__ult-=1
        return temp

    def recorrer(self):
        i = self.__pri
        while i < self.__ult:
            print(self.__cola[i], end=" ")
            i += 1
        print()

    def vacia(self):
        return self.__pri == self.__ult

    def llena(self):
        return self.__ult == self.__cap


if __name__ == "__main__":
    cola = Cola(3)
    cola.insertar(1)
    cola.insertar(2)
    cola.insertar(3)
    cola.insertar(4)

    cola.recorrer()

    cola.eliminar()

    cola.recorrer()
